WordPress/openverse
# Update VCollectionHeader.vue

## Fixes



Updated implementation of #2729 by @obulat

## Description


This PR is extracted from #3140 to make them easier to review.

When adding Nuxt page for the tag/source/creator collections in #3140, I first tried to add a single dynamic page for all of them to prevent code duplication.

I used a dynamic catch-all path `_mediaType/_collection/_.vue`, and thought that `_mediaType` would catch the `image` or `audio` path parameter, `_collection` would catch the first part of the path (`tag` or `source`),  and `_.vue` would be the catch-all that collects the rest of the path parameters: `tag/tagName`, `source/sourceName`, `source/sourceName/creator/creatorName`.

However, this strategy catches more routes than expected. For instance, it would catch `/<localePrefix>/non-existent-page`, and would return the English error page because the collection page was seeing the locale prefix as `_mediaType`, and path validation would fail because the locale prefix does not match the `_mediaType` (image/audio).

This is why I added 4 pages that used less dynamic path parameters (`/image/tag/_tag.vue`, `/image/source/_.vue`).
Previously, I planned to use the page to compute all the necessary parameters, and have `VCollectionHeader` as a simple presentational component. However, now that there are 4 page component, moving the computation to the store and the component makes more sense. 

Now `VCollectionHeader` accepts the `collectionParams` prop from the page component, and then computes the title, URL and the results label. There is a complication for the creator pages: we can only get the creator URL after we fetch the media, from the media item's `creator_url` param. This is why the `creatorUrl` is still computed in the page (which handles media fetching) and is passed as a prop to `collectionParams`.

This PR also converts the `CollectionParams` type to a [discriminated union type](https://basarat.gitbook.io/typescript/type-system/discriminated-unions) to make it easier to narrow the type when handling it.

## Testing Instructions

Check the Storybook to see that `VCollectionHeader` is rendered correctly: https://docs.openverse.org/_preview/3365/storybook/?path=/story/components-vcollectionheader--all-collections
To run the Storybook locally, you may need to run `just frontend/run test:storybook` to run Storybook inside docker (it currently does not run locally on Mac due to ssl errors).

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
